docs/architecture/adrs/005-chat-completion-support.md

```diff
@@ -0,0 +1,74 @@
+- For output detectors that run on the contents of an entire output stream, results will be included as `detections` on the second-last event (last one before `[DONE]`) of the stream, regardless of presence of `usage`.
+    - An alternate consideration made here was the currently implemented aggregation strategy, where if the user requested any output detectors that required the entire output, the stream events with detection results would only be returned all together. The problem here was the results would then not really be "streamed" back.
+
+
+#### User interactions
+
+This part gives a brief overview of how we would expect a user to access information from the response object.
+- Identifying which choice of potentially multiple choices that results exist for: `detections.output[0].choice_index` (will give index of choice in original `choices`)
+- Result info of a given choice: `detections.output[0].results[1].detection_type`
+
+### New elements
+- We will add a `warnings` element to show cases where the response is not 4xx, but there are issues in processing.
+
+## Consequences
+- Only additive changes to the OpenAI chat completions results will be made in the unary case. In the streaming case, streamed events may need to have updated text in `delta.content`. This can affect token counts per event in `usage`.
+- Validation and error handling will have to be provided for the cases where there is not expected `content` to run detectors on, raising warnings to the user when appropriate.
+- Since the results of various detectors can be included as the results of each `choice`, the inclusion of `detector_id` in each result will help distinguish which detector produced which result.
+- Results from input detectors will include `message_index` to indicate which message was processed. In cases where only the last message of the chat history may be processed to limit resource usage, this will help make it obvious to the user which message was processed.
+- Similarly, since output detectors will be run on all `choices` provided by the chat completions model, `choice_index` will be provided to inform the user which `choice` that detector result(s) apply for. This will apply regardless if the detector itself may take full or partial input to chat completions model.
+- A new result aggregation strategy will need to be implemented to put any detector results that require the full output text from a chat completions model (e.g. all the text of a particular `choice`) on the second-last event.
+- For the streaming response, implementation-wise, aggregating the streaming responses will be complicated since currently for the openAI chat completions API, each stream event returns content for only one choice, e.g.
```

Comment:
Shall we also add a point that each of the detector type will implement a "filtering" mechanism to select the messages they can work with ? 🤔

Reply:
yep, that sounds more accurate.